kotlin/workflow-testing/src/test/java/com/squareup/workflow/WorkerTest.kt

```diff
@@ -0,0 +1,243 @@
+/*
+ * Copyright 2019 Square Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+@file:Suppress("EXPERIMENTAL_API_USAGE")
+
+package com.squareup.workflow
+
+import com.squareup.workflow.testing.test
+import kotlinx.coroutines.Job
+import kotlinx.coroutines.channels.Channel
+import kotlinx.coroutines.channels.produce
+import kotlin.test.Test
+import kotlin.test.assertEquals
+import kotlin.test.assertFalse
+import kotlin.test.assertNotNull
+import kotlin.test.assertNotSame
+import kotlin.test.assertTrue
+
+class WorkerTest {
+
+  private class ExpectedException : RuntimeException()
+
+  @Test fun `create returns equivalent workers unkeyed`() {
+    val worker1 = Worker.create<Unit> {}
+    val worker2 = Worker.create<Unit> {}
+
+    assertNotSame(worker1, worker2)
+    assertTrue(worker1.doesSameWorkAs(worker2))
+  }
+
+  @Test fun `create returns equivalent workers keyed`() {
+    val worker1 = Worker.create<Unit>("key") {}
+    val worker2 = Worker.create<Unit>("key") {}
+
+    assertNotSame(worker1, worker2)
+    assertTrue(worker1.doesSameWorkAs(worker2))
+  }
+
+  @Test fun `create returns non-equivalent workers based on type`() {
+    val worker1 = Worker.create<Unit> {}
+    val worker2 = Worker.create<Int> {}
+
+    assertFalse(worker1.doesSameWorkAs(worker2))
+  }
+
+  @Test fun `create returns non-equivalent workers based on key`() {
+    val worker1 = Worker.create<Unit>("key1") {}
+    val worker2 = Worker.create<Unit>("key2") {}
+
+    assertFalse(worker1.doesSameWorkAs(worker2))
+  }
+
+  @Test fun `create emits and finishes`() {
+    val worker = Worker.create<String> {
```

Comment:
Fun fact: If we add the `@BuilderInference` annotation to this lambda parameter, the compiler can infer `String` from the `emitOutput` calls below and we can remove all these explicit type arguments. The IDE will even encourage us to do so.

It's _extremely_ tempting, but `@BuilderInference` and all its new type inference friends are still marked experimental so I'm resisting for now. That said, there's no explicit API here, and most real-world usages of workers in our codebase will use the Rx extensions, so the blast radius would probably be pretty small…

Reply:
Don't you need to use the experimental type inference for that? The last time I've tested it many things broke, especially the Rx calls.

Reply:
Yea.

Reply:
> Don't you need to use the experimental type inference for that?

Yes.

> The last time I've tested it many things broke, especially the Rx calls.

😢